File: py/creational/abstract_factory.py
# Abstract Factory creates related objects without naming their concrete classes.
# Modern Python note: this can often be a small factory function or config map.
# Keep the formal shape when related products must be created consistently.
from abc import ABC, abstractmethod


class Button(ABC):
    @abstractmethod
    def render(self):
        pass


class Checkbox(ABC):
    @abstractmethod
    def render(self):
        pass


class LightButton(Button):
    def render(self):
        return "light button"


class LightCheckbox(Checkbox):
    def render(self):
        return "light checkbox"


class DarkButton(Button):
    def render(self):
        return "dark button"


class DarkCheckbox(Checkbox):
    def render(self):
        return "dark checkbox"


class WidgetFactory(ABC):
    @abstractmethod
    def create_button(self):
        pass

    @abstractmethod
    def create_checkbox(self):
        pass


class LightFactory(WidgetFactory):
    def create_button(self):
        return LightButton()

    def create_checkbox(self):
        return LightCheckbox()


class DarkFactory(WidgetFactory):
    def create_button(self):
        return DarkButton()

    def create_checkbox(self):
        return DarkCheckbox()


def build_screen(factory):
    return [factory.create_button().render(), factory.create_checkbox().render()]


if __name__ == "__main__":
    print(build_screen(DarkFactory()))

File: csharp/creational/abstract_factory.cs
namespace Gof.ModernCSharp.Creational;

// Abstract Factory creates related objects without naming their concrete classes.
// Modern C# note: this is often a small factory service registered in DI. Keep
// the formal shape when related products must be created consistently.
public static class AbstractFactory
{
    interface IButton { string Render(); }
    interface ICheckbox { string Render(); }
    interface IWidgetFactory
    {
        IButton CreateButton();
        ICheckbox CreateCheckbox();
    }

    sealed class DarkButton : IButton { public string Render() => "dark button"; }
    sealed class DarkCheckbox : ICheckbox { public string Render() => "dark checkbox"; }
    sealed class DarkFactory : IWidgetFactory
    {
        public IButton CreateButton() => new DarkButton();
        public ICheckbox CreateCheckbox() => new DarkCheckbox();
    }

    public static string Run()
    {
        IWidgetFactory factory = new DarkFactory();
        return $"{factory.CreateButton().Render()}, {factory.CreateCheckbox().Render()}";
    }
}
